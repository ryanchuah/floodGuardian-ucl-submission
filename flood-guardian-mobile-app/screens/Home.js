import React, {useState, useEffect} from 'react';
import {View, Text, StyleSheet, TouchableOpacity} from 'react-native';
import 'react-native-get-random-values';
import LinearGradient from 'react-native-linear-gradient';
import Icon from 'react-native-vector-icons/MaterialIcons';

import Header from '../components/Header';
import Forecast from '../components/home/Forecast';
import LocationNotSet from '../components/home/LocationNotSet';
import LocationModal from '../components/home/LocationModal.js';
import {getData} from '../helper';

const Home = () => {
  const [modalVisible, setModalVisible] = useState(false);
  const [locationHistory, setLocationHistory] = useState(null);
  const [lastUpdated, setLastUpdated] = useState('Loading...');
  const [currLat, setCurrLat] = useState(null);
  const [currLong, setCurrLong] = useState(null);
  const [coordsLoading, setCoordsLoading] = useState({lat: true, long: true});
  const [locationHistLoading, setLocationHistLoading] = useState(true);
  useEffect(() => {
    getData('locationHistory').then((hist) => {
      if (hist) setLocationHistory(hist);
      setLocationHistLoading(false);
    });
    getData('lastUpdated').then((time) => {
      if (time) setLastUpdated(time);
    });
    getData('currLat').then((lat) => {
      if (lat) setCurrLat(lat);
      setCoordsLoading((prevState) => ({...prevState, lat: false}));
    });
    getData('currLong').then((long) => {
      if (long) setCurrLong(long);
      setCoordsLoading((prevState) => ({...prevState, long: false}));
    });
  }, [(currLat, currLong)]);

  return (
    <LinearGradient
      colors={['#2C659A', 'rgba(0, 0, 0, 0)']}
      style={styles.linearGradient}>
      <Header />

      <View style={styles.location}>
        <TouchableOpacity
          style={styles.locationContainer}
          onPress={() => {
            setModalVisible(true);
          }}>
          <View style={styles.locationIcon}>
            <Icon name="location-pin" color="white" size={40} />
          </View>
          <View style={styles.locationTextContainer}>
            <Text style={styles.locationText}>
              {locationHistory ? locationHistory[0] : 'Tap to set location'}
            </Text>
          </View>
        </TouchableOpacity>
      </View>

      <View style={styles.forecast}>
        {locationHistory || locationHistLoading ? (
          <>
            <Text
              style={styles.lastUpdated}>{`Last Updated: ${lastUpdated}`}</Text>
            <Forecast
              key={(currLat, currLong)}
              setLastUpdated={setLastUpdated}
              currLat={currLat}
              currLong={currLong}
              coordsLoading={coordsLoading}
              locationHistLoading={locationHistLoading}
            />
          </>
        ) : (
          <LocationNotSet />
        )}
      </View>

      <View style={styles.modalCenteredView}>
        <LocationModal
          setModalVisible={setModalVisible}
          modalVisible={modalVisible}
          setLocationHistory={setLocationHistory}
          locationHistory={locationHistory}
          setCurrLat={setCurrLat}
          setCurrLong={setCurrLong}
          // setReloadHome={setReloadHome}
        />
      </View>
    </LinearGradient>
  );
};

const styles = StyleSheet.create({
  forecast: {
    alignSelf: 'center',
    backgroundColor: 'rgba(38, 69, 148, 0.08)',
    width: 350,
    borderRadius: 10,
    marginBottom: 15,
  },
  lastUpdated: {
    textAlign: 'right',
    color: '#fff',
    paddingTop: 8,
    paddingBottom: 4,
    paddingRight: 15,
  },
  location: {
    flexDirection: 'row',
    justifyContent: 'center',
    marginBottom: 15,
  },
  locationContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 15,
    marginLeft: 70,
    marginRight: 70,
  },
  locationIcon: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  locationTextContainer: {
    justifyContent: 'flex-end',
  },
  locationText: {
    color: 'white',
    fontSize: 25,
    fontFamily: 'Roboto-Regular',
    fontWeight: 'bold',
    paddingLeft: 5,
    textAlign: 'center',
  },
  linearGradient: {
    flex: 1,
  },
});
export default Home;
