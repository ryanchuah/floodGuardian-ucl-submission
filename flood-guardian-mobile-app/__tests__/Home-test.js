import React from 'react';
import renderer from 'react-test-renderer';
import Home from '../screens/Home';

jest.mock('react-native-geolocation-service', () => ({
    getCurrentPosition: jest.fn()
  }));

test('home page renders correctly', async () => {
  const tree = await renderer.create(<Home />).toJSON();
  expect(tree).toMatchSnapshot();
});