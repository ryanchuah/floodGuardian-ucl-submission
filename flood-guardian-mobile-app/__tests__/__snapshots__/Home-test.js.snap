// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home page renders correctly 1`] = `
<BVLinearGradient
  colors={
    Array [
      4281099674,
      0,
    ]
  }
  endPoint={
    Object {
      "x": 0.5,
      "y": 1,
    }
  }
  locations={null}
  startPoint={
    Object {
      "x": 0.5,
      "y": 0,
    }
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "paddingBottom": 20,
        "paddingTop": 30,
      }
    }
  >
    <Text
      style={
        Object {
          "alignItems": "center",
          "color": "#9BBFE9",
          "flex": 1,
          "fontFamily": "Chicle-Regular",
          "fontSize": 50,
          "justifyContent": "center",
          "textAlign": "center",
        }
      }
    >
      Flood Guardian
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 15,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 15,
          "marginLeft": 70,
          "marginRight": 70,
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 40,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
      <View
        style={
          Object {
            "justifyContent": "flex-end",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontFamily": "Roboto-Regular",
              "fontSize": 25,
              "fontWeight": "bold",
              "paddingLeft": 5,
              "textAlign": "center",
            }
          }
        >
          Tap to set location
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "rgba(38, 69, 148, 0.08)",
        "borderRadius": 10,
        "marginBottom": 15,
        "width": 350,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "paddingBottom": 4,
          "paddingRight": 15,
          "paddingTop": 8,
          "textAlign": "right",
        }
      }
    >
      Last Updated: Loading...
    </Text>
    <View
      style={
        Object {
          "alignContent": "center",
          "alignSelf": "center",
          "height": 200,
          "justifyContent": "center",
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#000"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
  <View>
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      transparent={true}
      visible={false}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "marginTop": 22,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "elevation": 5,
              "height": null,
              "width": 350,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "stretch",
                "color": "#4d4d4d",
                "fontSize": 20,
                "paddingBottom": 30,
                "paddingLeft": 30,
                "paddingTop": 20,
                "textAlign": "left",
              }
            }
          >
            Location
          </Text>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "opacity": 1,
                "paddingBottom": 15,
                "paddingLeft": 10,
                "paddingRight": 10,
                "width": 250,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#2696FF",
                    "fontSize": 40,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
            <Text
              style={
                Object {
                  "fontSize": 20,
                  "marginHorizontal": 10,
                  "marginLeft": 10,
                }
              }
            >
              My current location
            </Text>
          </View>
          <View
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#2696FF",
                  "fontSize": 20,
                  "fontWeight": "bold",
                  "paddingBottom": 10,
                  "paddingTop": 10,
                }
              }
            >
              Add a location
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2696FF",
                    "fontSize": 20,
                    "fontWeight": "bold",
                    "paddingBottom": 10,
                    "paddingTop": 10,
                  },
                  Object {
                    "paddingBottom": 30,
                  },
                ]
              }
            >
              Close
            </Text>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</BVLinearGradient>
`;
